GUVNOR-3474: grouping columns in management section

jomarko opened the pull request:
Missing tests, and probably I will change grouping of DRL columns to have all of them together.

manstis commented:
@karreiro Hi, would you mind taking a look for @jomarko

jomarko commented:
![screenshot from 2017-09-12 14-36-30](https://user-images.githubusercontent.com/8044780/30326132-cc4e469c-97c7-11e7-8a7d-37962526a1fd.png)
![screenshot from 2017-09-12 14-36-23](https://user-images.githubusercontent.com/8044780/30326134-cc6d7594-97c7-11e7-91be-43cb0206e540.png)

These pictures shows grouping of columns in the management section.

karreiro commented on drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/pages/common/DecisionTableColumnViewUtils.java:
```diff
@@ -60,6 +77,51 @@ public static void addWidgetToContainer(final IsWidget widget,
+        } else if (column instanceof ActionWorkItemSetFieldCol52) {
+            managementSectionHeader = concatenateFactTypeAndBoundName(null,
+                                                                      ((ActionWorkItemSetFieldCol52) column).getBoundName());
+        } else if (column instanceof Pattern52) {
```
I think that we could remove the `if`s related to `ActionWorkItemInsertFactCol52` and `ActionWorkItemSetFieldCol52`, since they extends respectively `ActionInsertFactCol52` and `ActionSetFieldCol52`.

jomarko commented on drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTableModellerViewImpl.java:
```diff
@@ -466,131 +470,25 @@ private SmallLabel makeColumnLabel(final MetadataCol52 metaDataColumn) {
+        final Map<String, List<BaseColumn>> columnGroups =
+                conditionColumns.stream().collect(
+                        Collectors.groupingBy(
+                                column -> DecisionTableColumnViewUtils.getColumnManagementGroupTitle(column),
```
Nice hint, fixed.

jomarko commented on drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTableModellerViewImpl.java:
```diff
@@ -606,94 +504,28 @@ private Anchor anchor(final String text,
+        final Map<String, List<BaseColumn>> columnGroups =
+                actionColumns.stream().collect(
+                        Collectors.groupingBy(
+                                column -> DecisionTableColumnViewUtils.getColumnManagementGroupTitle(column),
```
Fixed

jomarko commented on drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTableModellerViewImpl.java:
```diff
@@ -850,6 +682,20 @@ Transform newTransform() {
+        if (actionsPanel == null) {
+            actionsPanel = new ColumnManagementView(getPresenter());
+        }
+        return actionsPanel;
```
Nice, I like this way, fixed.

jomarko commented on drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTableModellerViewImpl.java:
```diff
@@ -850,6 +682,20 @@ Transform newTransform() {
+    }
+
+    ColumnManagementView getConditionsPanel() {
+        if (conditionsPanel == null) {
```
Fixed.

jomarko commented on drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/accordion/ColumnManagementView.java:
```diff
@@ -0,0 +1,254 @@
+                } else if (column instanceof BRLConditionColumn) {
+                    columnGroupColumns.add(renderColumn((BRLConditionColumn) column));
+                } else if (column instanceof Pattern52) {
+                    renderColumn((Pattern52) column).forEach(renderedCondition -> columnGroupColumns.add(renderedCondition));
```
Fixed

jomarko commented on drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/accordion/ColumnManagementView.java:
```diff
@@ -0,0 +1,254 @@
+        return label;
+    }
+
+    SmallLabel makeColumnLabel(final ActionCol52 actionColumn) {
```
Covered

jomarko commented on drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/accordion/ColumnManagementView.java:
```diff
@@ -0,0 +1,254 @@
+        return condition;
+    }
+
+    SmallLabel makeColumnLabel(final ConditionCol52 cc) {
```
Covered

jomarko commented on drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/accordion/ColumnManagementView.java:
```diff
@@ -0,0 +1,254 @@
+        return deleteAnchor(clickHandler);
+    }
+
+    SmallLabel makeColumnLabel(final BRLConditionColumn brl) {
```
I removed this method, `SmallLabel makeColumnLabel(final ConditionCol52 conditionColumn)` is used now both for BRL and simple conditions.

jomarko commented on drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/pages/common/DecisionTableColumnViewUtils.java:
```diff
@@ -16,8 +16,25 @@
+import org.drools.workbench.models.guided.dtable.shared.model.CompositeColumn;
+import org.drools.workbench.models.guided.dtable.shared.model.Pattern52;
+import org.drools.workbench.screens.guided.dtable.client.resources.i18n.GuidedDecisionTableConstants;
+import org.drools.workbench.services.verifier.api.client.index.Pattern;
```
Fixed

jomarko commented on drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/accordion/ColumnManagementViewTest.java:
```diff
@@ -0,0 +1,270 @@
+
+import com.google.gwt.event.dom.client.ClickHandler;
+import com.google.gwt.user.client.ui.HorizontalPanel;
+import com.google.gwt.user.client.ui.VerticalPanel;
```
Fixed

jomarko commented on drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTableModellerViewImplTest.java:
```diff
@@ -34,9 +36,24 @@
+import org.drools.workbench.models.guided.dtable.shared.model.BRLConditionColumn;
+import org.drools.workbench.models.guided.dtable.shared.model.BaseColumn;
+import org.drools.workbench.models.guided.dtable.shared.model.CompositeColumn;
+import org.drools.workbench.models.guided.dtable.shared.model.ConditionCol52;
```
Fixed

jomarko commented on drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/accordion/ColumnManagementView.java:
```diff
@@ -0,0 +1,254 @@
+        if (column instanceof BRLConditionColumn) {
+            return makeRemoveConditionWidget(column,
+                                             (command) -> {
+                                                 if (!presenter.getActiveDecisionTable().canConditionBeDeleted((BRLConditionColumn) column)) {
```
Methods creating widget for removing simplified and covered by tests.

jomarko commented on drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/accordion/ColumnManagementView.java:
```diff
@@ -0,0 +1,254 @@
+
+        return makeRemoveConditionWidget(column,
+                                         (command) -> {
+                                             if (!presenter.getActiveDecisionTable().canConditionBeDeleted(column)) {
```
Same as above.

jomarko commented:
@karreiro thanks for this thorough review. I updated the PR, please notice that I also updated style of hidden columns, see pictures below, However now the ITALIC is not applied, just the color, do you have any experience with something similar?
Hidden columns before:
![screenshot from 2017-09-13 11-35-04](https://user-images.githubusercontent.com/8044780/30370448-ee322d18-9877-11e7-8596-f55f624323be.png)
Hidden columns after:
![screenshot from 2017-09-13 11-33-04](https://user-images.githubusercontent.com/8044780/30370457-f698c1ba-9877-11e7-8c62-02a54ef5b52c.png)

karreiro commented:
Hi @jomarko,

I think that you could fix the italic thing by adding something like:
```css
font-style: italic;
```
here: https://github.com/kiegroup/drools-wb/blob/master/drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/accordion/GuidedDecisionTableAccordionItemView.less#L41

karreiro commented:
@jomarko Unfortunately, I think that the hidden and non-hidden columns are _italic_, now. See:

![italic](https://user-images.githubusercontent.com/1079279/30502311-16a04044-9a3c-11e7-8311-80254279f244.gif)

jomarko commented:
@karreiro I think I got it now, please have a look https://webmshare.com/aogrM

karreiro commented:
Thanks, @jomarko. Looks nice now :-)

I think that it's just a matter of fixing the conflicts and we're ready to roll.

jomarko commented:
@karreiro conflicts resolved, lets see if tests pass.
